amlib/dsp/stereo_convolution_mac.py

```diff
@@ -0,0 +1,439 @@
+                                ]
+                            elif self._convolutionMode == ConvolutionMode.MONO:
+                                m.d.sync += [
+                                    madd_values[i].eq(madd_values[i] + (left_sample * main_tap)),
+                                    madd_values[i + self._slices].eq(madd_values[i + self._slices] + (right_sample * main_tap)),
+                                ]
+
+                # shift the samples buffer by one sample to prepare for the next arriving sample in the IDLE state
+                with m.If(ix > 1):
+                    m.d.sync += [
+                        samples1_write_port.data.eq(Cat((previous_sample1 >> self._bitwidth)[:((self._slices - 1) * self._bitwidth)], carryover1_2)),
+                        samples1_write_port.addr.eq(ix - 2),
+                        samples1_write_port.en.eq(1),
+
+                        samples2_write_port.data.eq(Cat((previous_sample2 >> self._bitwidth)[:((self._slices - 1) * self._bitwidth)], carryover2_2)),
+                        samples2_write_port.addr.eq(ix - 2),
+                        samples2_write_port.en.eq(1),
+                    ]
+
+                with m.If(ix == self._samples_per_slice + 1):
+                    m.next = "SUM"
+                with m.Else():
+                    m.d.sync += ix.eq(ix+1)
+
+            with m.State("SUM"):
+                sumL = 0
+                sumR = 0
+                for i in range(self._slices):
+                    sumL += madd_values[i]
+                    sumR += madd_values[i + self._slices]
+
+                m.d.sync += [
+                    sumSignalL.eq(sumL),
+                    sumSignalR.eq(sumR),
+                    output_channels.eq(0),
+                ]
+                m.next = "OUTPUT"
+            with m.State("OUTPUT"):
+                m.d.sync += [
+                    set1.eq(0),
+                    set2.eq(0),
+                    ix.eq(0),
+                ]
+
+                with m.If(output_channels == 2):
+                    m.next = "IDLE"
+                with m.Elif(self.signal_out.ready):
+                    m.d.sync += [
+                        output_channels.eq(output_channels + 1),
+                        self.signal_out.payload.eq(Mux(output_channels == 0, sumSignalL >> self._bitwidth, sumSignalR >> self._bitwidth)),
+                        self.signal_out.valid.eq(1),
+                        self.signal_out.first.eq(~output_channels),
+                        self.signal_out.last.eq(output_channels),
+                    ]
+
+        return m
+
+
+class StereoConvolutionMACTest(GatewareTestCase):
+    FRAGMENT_UNDER_TEST = StereoConvolutionMAC
+    testSamplecount = 120
+    tapcount = 32
+    bitwidth = 24
+    samplerate = 48000
+    clockfrequency = samplerate * tapcount / 4 # we want to test for 4 slices
+
+    #some test IR-data
```

Comment:
Is this still necessary, and if yes, for what. Please delete otherwise. Thanks!

Reply:
Yes, it is passed in the FRAGMENT_ARGUMENTS.
The logic calculates how many clock cycles are available in the sync domain per signal sample.
E.g. 60MHz sync domain and 48kHz samplerate leads to 1250 cycles available for calculations per audio sample. If you want to process an IR with 4096 taps you need 4 parallel computations (slices) in order to do real-time dsp.

Since for testing it is unrealistic to process 4000 taps we specify a slow clockfrequency of 384kHz in order to test the calculation for 4 slices. Since testing for a single slice would not uncover any parallel-processing bugs of the code.

Reply:
You need to put this in the comment why it is necessary. We need that in the code.

Reply:
I meant the commented out test IR-data (tapdata1 and tapdata2 below).
I have no idea what the commented out code is good for.
Please comment why it is necessary or remove.

Reply:
Thanks for the additional comments. That will really help the reader of the code!